packages/datetime/src/dateRangeSelectionStrategy.ts

```diff
@@ -0,0 +1,155 @@
+
+import {
+    areSameDay,
+    DateRange,
+    DateRangeBoundary,
+} from "./common/dateUtils";
+
+export interface IDateRangeSelectionState {
+    /**
+     * The boundary that would be modified by clicking the provided `day`.
+     * May be different from `boundaryToModify` in some special cases
+     * (e.g. when hovering over the other boundary's selected date).
+     */
+    boundary?: DateRangeBoundary;
+
+    /**
+     * The date range that would be selected after clicking the provided `day`.
+     */
+    dateRange: DateRange;
+};
+
+export class DateRangeSelectionStrategy {
+    public static getNextState(currentRange: DateRange,
+                               day: Date,
+                               boundaryToModify?: DateRangeBoundary,
+                               allowSingleDayRange?: boolean) {
+        if (boundaryToModify != null) {
+            return this.getNextStateForBoundary(currentRange, day, boundaryToModify, allowSingleDayRange);
+        } else {
+            return this.getDefaultNextState(currentRange, day, allowSingleDayRange);
+        }
+    }
+
+    private static getNextStateForBoundary(currentRange: DateRange,
+                                           day: Date,
+                                           boundaryToModify: DateRangeBoundary,
+                                           allowSingleDayRange: boolean) {
+        const boundary = boundaryToModify;
+        const boundaryDate = this.getBoundaryDate(boundary, currentRange);
+
+        const otherBoundary = this.getOtherBoundary(boundary);
+        const otherBoundaryDate = this.getBoundaryDate(otherBoundary, currentRange);
+
+        let nextBoundary: DateRangeBoundary;
+        let nextDateRange: DateRange;
+
+        if (boundaryDate == null && otherBoundaryDate == null) {
+            nextBoundary = boundary;
+            nextDateRange = this.createRangeForBoundary(boundary, day, null);
+        } else if (boundaryDate != null && otherBoundaryDate == null) {
+            const nextBoundaryDate = areSameDay(boundaryDate, day) ? null : day;
+            nextBoundary = boundary;
+            nextDateRange = this.createRangeForBoundary(boundary, nextBoundaryDate, null);
+        } else if (boundaryDate == null && otherBoundaryDate != null) {
+            if (areSameDay(day, otherBoundaryDate)) {
+                const [nextBoundaryDate, nextOtherBoundaryDate] = (allowSingleDayRange)
+                    ? [otherBoundaryDate, otherBoundaryDate]
+                    : [null, null];
+                nextBoundary = allowSingleDayRange ? boundary : otherBoundary;
```

Comment:
Tried that earlier and had trouble, because `nextBoundary` needs to _not_ use `const` (it's mutable and already declared), yet `nextBoundaryDate` and `nextOtherBoundaryDate` should use `const` within this scope. Is there a reasonable way around that?